File: app/lib/gamification/badge_engine.dart
final _badgeConditions = <String, bool Function(Map<String, dynamic>)>{
  'FIRST_WORKOUT': (s) => (s['totalWorkoutsCompleted'] as int) >= 1,
  'ONE_WEEK_STREAK': (s) => (s['longestStreak'] as int) >= 7,
  'HUNDRED_DAYS': (s) => (s['longestStreak'] as int) >= 100,
  'IRON_STOMACH': (s) => (s['totalDaysAllMealsCompleted'] as int) >= 7,
  'FITNESS_OWL': (s) => (s['level'] as int) >= 10,
};

List<String> checkUnlockedBadges(Map<String, dynamic> stats) {
  return _badgeConditions.entries
      .where((e) => e.value(stats))
      .map((e) => e.key)
      .toList();
}

File: app/android/app/src/main/java/com/gymcoach/gym_coach/gamification/BadgeEngine.kt
package com.gymcoach.gym_coach.gamification

/**
 * Mirrors `lib/gamification/badge_engine.dart`. The function is a pure
 * predicate over a stats map; no DB lookup.
 *
 * Required stats:
 *   totalWorkoutsCompleted (Int)
 *   longestStreak         (Int)
 *   totalDaysAllMealsCompleted (Int)
 *   level                 (Int)
 */
object BadgeEngine {

    const val FIRST_WORKOUT  = "FIRST_WORKOUT"
    const val ONE_WEEK_STREAK = "ONE_WEEK_STREAK"
    const val HUNDRED_DAYS = "HUNDRED_DAYS"
    const val IRON_STOMACH = "IRON_STOMACH"
    const val FITNESS_OWL = "FITNESS_OWL"

    fun checkUnlockedBadges(stats: Map<String, Int>): List<String> {
        val totalWorkouts = stats.getOrDefault("totalWorkoutsCompleted", 0)
        val longestStreak = stats.getOrDefault("longestStreak", 0)
        val allMealsDays  = stats.getOrDefault("totalDaysAllMealsCompleted", 0)
        val level         = stats.getOrDefault("level", 0)
        val out = mutableListOf<String>()
        if (totalWorkouts >= 1) out += FIRST_WORKOUT
        if (longestStreak >= 7) out += ONE_WEEK_STREAK
        if (longestStreak >= 100) out += HUNDRED_DAYS
        if (allMealsDays >= 7) out += IRON_STOMACH
        if (level >= 10) out += FITNESS_OWL
        return out
    }
}
